File: solo-leveling-player-service/src/main/java/com/sleepkqq/sololeveling/player/service/mapper/AvroMapper.java
package com.sleepkqq.sololeveling.player.service.mapper;

import com.sleepkqq.sololeveling.avro.task.SaveTask;
import com.sleepkqq.sololeveling.player.service.model.task.Task;
import com.sleepkqq.sololeveling.player.service.model.task.TaskDraft;
import com.sleepkqq.sololeveling.player.service.model.task.enums.TaskRarity;
import com.sleepkqq.sololeveling.player.service.model.task.enums.TaskTopic;
import org.springframework.stereotype.Component;

@Component
public class AvroMapper extends BaseMapper {

  public Task map(SaveTask saveTask) {
    return new TaskDraft.Builder()
        .id(map(saveTask.getTaskId()))
        .title(saveTask.getTitle())
        .description(saveTask.getDescription())
        .experience(saveTask.getExperience())
        .rarity(map(saveTask.getRarity()))
        .topics(mapCollection(saveTask.getTopics(), this::map))
        .agility(saveTask.getAgility())
        .strength(saveTask.getStrength())
        .intelligence(saveTask.getIntelligence())
        .build();
  }

  public TaskRarity map(com.sleepkqq.sololeveling.avro.task.TaskRarity taskRarity) {
    if (taskRarity == null) {
      return null;
    }
    return TaskRarity.valueOf(taskRarity.name());
  }

  public TaskTopic map(com.sleepkqq.sololeveling.avro.task.TaskTopic topic) {
    if (topic == null) {
      return null;
    }
    return TaskTopic.valueOf(topic.name());
  }

  public com.sleepkqq.sololeveling.avro.task.TaskRarity map(TaskRarity taskRarity) {
    if (taskRarity == null) {
      return null;
    }
    return com.sleepkqq.sololeveling.avro.task.TaskRarity.valueOf(taskRarity.name());
  }

  public com.sleepkqq.sololeveling.avro.task.TaskTopic map(TaskTopic topic) {
    if (topic == null) {
      return null;
    }
    return com.sleepkqq.sololeveling.avro.task.TaskTopic.valueOf(topic.name());
  }
}

File: solo-leveling-player-service/src/main/kotlin/com/sleepkqq/sololeveling/player/service/mapper/AvroMapper.kt
package com.sleepkqq.sololeveling.player.service.mapper

import com.sleepkqq.sololeveling.avro.task.GenerateTask
import com.sleepkqq.sololeveling.avro.task.SaveTask
import com.sleepkqq.sololeveling.player.model.entity.task.Task
import com.sleepkqq.sololeveling.player.model.entity.task.dto.TaskInput
import com.sleepkqq.sololeveling.player.model.entity.task.enums.TaskRarity
import com.sleepkqq.sololeveling.player.model.entity.task.enums.TaskTopic
import org.mapstruct.CollectionMappingStrategy
import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.Named
import org.mapstruct.ReportingPolicy
import org.springframework.stereotype.Component
import java.util.UUID

@Component
@Mapper(
	componentModel = "spring",
	unmappedTargetPolicy = ReportingPolicy.IGNORE,
	collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED
)
abstract class AvroMapper {

	@Named("toEntityTaskRarity")
	fun map(taskRarity: com.sleepkqq.sololeveling.avro.task.TaskRarity): TaskRarity =
		TaskRarity.valueOf(taskRarity.name)

	@Named("toAvroTaskRarity")
	fun map(taskRarity: TaskRarity): com.sleepkqq.sololeveling.avro.task.TaskRarity =
		com.sleepkqq.sololeveling.avro.task.TaskRarity.valueOf(taskRarity.name)

	@Named("toEntityTaskTopic")
	fun map(topic: com.sleepkqq.sololeveling.avro.task.TaskTopic): TaskTopic =
		TaskTopic.valueOf(topic.name)

	@Named("toAvroTaskTopic")
	fun map(topic: TaskTopic): com.sleepkqq.sololeveling.avro.task.TaskTopic =
		com.sleepkqq.sololeveling.avro.task.TaskTopic.valueOf(topic.name)

	fun map(saveTask: SaveTask): TaskInput = TaskInput(
		id = UUID.fromString(saveTask.taskId),
		title = saveTask.title,
		description = saveTask.description,
		experience = saveTask.experience,
		currencyReward = saveTask.currencyReward,
		rarity = map(saveTask.rarity),
		topics = saveTask.topics.map { map(it) },
		agility = saveTask.agility,
		strength = saveTask.strength,
		intelligence = saveTask.intelligence,
		version = saveTask.version,
	)

	@Mapping(target = "taskId", source = "task.id")
	@Mapping(target = "topics", source = "topics", qualifiedByName = ["toAvroTaskTopic"])
	@Mapping(target = "rarity", source = "rarity", qualifiedByName = ["toAvroTaskRarity"])
	abstract fun map(task: Task, topics: Collection<TaskTopic>, rarity: TaskRarity): GenerateTask
}
